binary_search_tree/tree.py

```diff
@@ -1,59 +1,133 @@
 class TreeNode:
-    def __init__(self, key, val = None):
+    def __init__(self, key, val=None):
         if val == None:
             val = key
 
         self.key = key
         self.value = val
         self.left = None
         self.right = None
-        
 
 
 class Tree:
     def __init__(self):
         self.root = None
 
-    # Time Complexity: 
-    # Space Complexity: 
-    def add(self, key, value = None):
-        pass
+    def add_helper(self, current_node, new_node):
+        if current_node == None:
+            return new_node
+
+        if new_node.key <= current_node.key:
+            current_node.left = self.add_helper(current_node.left, new_node)
+        else:
+            current_node.right = self.add_helper(
+                current_node.right, new_node)
+        return current_node
+
+    # Time Complexity: O(log n)
+    # Space Complexity: O(1)
+    def add(self, key, value=None):
```

Comment:
👍 But the space complexity is O(log n) for a balanced tree due to recursive call stack.